rpc/client/httpclient.go

```diff
@@ -316,34 +309,27 @@ func (w *WSEvents) redoSubscriptions() {
 func (w *WSEvents) eventListener() {
 	for {
 		select {
-		case resp := <-w.ws.ResponsesCh:
-			// res is json.RawMessage
+		case resp, ok := <-w.ws.ResponsesCh:
+			if !ok {
+				return
+			}
 			if resp.Error != nil {
-				// FIXME: better logging/handling of errors??
-				fmt.Printf("ws err: %+v\n", resp.Error.Error())
+				w.Logger.Error("WS error", "err", resp.Error.Error())
 				continue
 			}
 			result := new(ctypes.ResultEvent)
 			err := json.Unmarshal(resp.Result, result)
 			if err != nil {
-				// ignore silently (eg. subscribe, unsubscribe and maybe other events)
-				// TODO: ?
+				w.Logger.Error("failed to unmarshal response", "err", err)
 				continue
 			}
-			if ch := w.getSubscription(result.Query); ch != nil {
+			w.mtx.RLock()
+			if ch, ok := w.subscriptions[result.Query]; ok {
 				ch <- result.Data
 			}
-		case <-w.quit:
-			// send a message so we can wait for the routine to exit
-			// before cleaning up the w.ws stuff
-			w.done <- true
+			w.mtx.RUnlock()
```

Comment:
probably better to move up a little bit (i.e. unblock before writing to channel)

Reply:
no. it is intended, so we can't close the channel

Reply:
wrote a comment